File: Backend/src/routers/relatorios.py
"""
Endpoints FastAPI para relatórios de execuções.
"""

from fastapi import APIRouter, HTTPException, status, Query
from typing import Optional, List
from datetime import datetime
from pydantic import BaseModel
import csv
from io import StringIO

from ..db.session import SessionLocal
from ..db.models import Execucao
from ..repositories.empresas_repo import get_empresa_by_id
from ..infrastructure.logger import get_logger

logger = get_logger(__name__)

router = APIRouter(prefix="/api/relatorios", tags=["Relatórios"])


class EmpresaResumo(BaseModel):
    """Resumo de uma empresa no relatório."""
    cnpj: str
    nome: Optional[str] = None
    qtd_notas_emitidas: int = 0
    qtd_notas_recebidas: int = 0


class ResumoExecucoesResponse(BaseModel):
    """Resposta do endpoint de resumo de execuções."""
    competencia: Optional[str]
    total_empresas: int
    com_movimento: int
    sem_movimento: int
    empresas_com_movimento: List[EmpresaResumo]
    empresas_sem_movimento: List[EmpresaResumo]


@router.get("/execucoes/resumo", response_model=ResumoExecucoesResponse, summary="Obter resumo das execuções")
def obter_resumo_execucoes(
    competencia: Optional[str] = Query(None, description="Competência no formato MMAAAA (ex: 112025). Se não informado, retorna todas."),
    status_filtro: Optional[str] = Query("concluido", description="Status para filtrar: concluido, falhou, etc. Padrão: concluido")
) -> ResumoExecucoesResponse:
    """
    Gera um resumo das execuções, classificando empresas em:
    - Com movimento: empresas que baixaram pelo menos uma nota (emitida ou recebida)
    - Sem movimento: empresas que não baixaram nenhuma nota
    
    Args:
        competencia: Competência no formato MMAAAA (opcional)
        status_filtro: Status das execuções para incluir no relatório (padrão: concluido)
        
    Returns:
        ResumoExecucoesResponse com estatísticas e listas de empresas
    """
    try:
        db = SessionLocal()
        try:
            # Monta query base
            query = db.query(Execucao).filter(Execucao.status == status_filtro)
            
            # Filtra por competência se fornecida
            if competencia:
                query = query.filter(Execucao.competencia == competencia)
            
            # Busca execuções
            execucoes = query.all()
            
            empresas_com_movimento = []
            empresas_sem_movimento = []
            
            for execucao in execucoes:
                # Determina se teve movimento
                qtd_emitidas = execucao.qtd_notas_emitidas or 0
                qtd_recebidas = execucao.qtd_notas_recebidas or 0
                
                # Busca nome da empresa se disponível
                nome_empresa = None
                if execucao.empresa_id:
                    try:
                        empresa = get_empresa_by_id(execucao.empresa_id)
                        if empresa:
                            nome_empresa = empresa.get("nome") or empresa.get("razao_social")
                    except:
                        pass
                
                empresa_resumo = EmpresaResumo(
                    cnpj=execucao.cnpj or execucao.empresa_id,
                    nome=nome_empresa,
                    qtd_notas_emitidas=qtd_emitidas,
                    qtd_notas_recebidas=qtd_recebidas
                )
                
                if qtd_emitidas > 0 or qtd_recebidas > 0:
                    empresas_com_movimento.append(empresa_resumo)
                else:
                    empresas_sem_movimento.append(empresa_resumo)
            
            return ResumoExecucoesResponse(
                competencia=competencia,
                total_empresas=len(execucoes),
                com_movimento=len(empresas_com_movimento),
                sem_movimento=len(empresas_sem_movimento),
                empresas_com_movimento=empresas_com_movimento,
                empresas_sem_movimento=empresas_sem_movimento
            )
            
        finally:
            db.close()
            
    except Exception as e:
        logger.error(f"Erro ao gerar resumo de execuções: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Erro ao gerar resumo: {str(e)}"
        )


@router.get("/execucoes/resumo/csv", summary="Baixar resumo das execuções em CSV")
def baixar_resumo_csv(
    competencia: Optional[str] = Query(None, description="Competência no formato MMAAAA (ex: 112025)"),
    status_filtro: Optional[str] = Query("concluido", description="Status para filtrar")
):
    """
    Gera e retorna um arquivo CSV com o resumo das execuções.
    
    Args:
        competencia: Competência no formato MMAAAA (opcional)
        status_filtro: Status das execuções para incluir no relatório
        
    Returns:
        Arquivo CSV para download
    """
    try:
        # Obtém resumo
        resumo = obter_resumo_execucoes(competencia=competencia, status_filtro=status_filtro)
        
        # Cria CSV em memória
        output = StringIO()
        writer = csv.writer(output)
        
        # Cabeçalho
        writer.writerow([
            "CNPJ",
            "Nome",
            "Total Notas Emitidas",
            "Total Notas Recebidas",
            "Total Notas",
            "Status"
        ])
        
        # Empresas com movimento
        for empresa in resumo.empresas_com_movimento:
            total_notas = empresa.qtd_notas_emitidas + empresa.qtd_notas_recebidas
            writer.writerow([
                empresa.cnpj,
                empresa.nome or "",
                empresa.qtd_notas_emitidas,
                empresa.qtd_notas_recebidas,
                total_notas,
                "Com movimento"
            ])
        
        # Empresas sem movimento
        for empresa in resumo.empresas_sem_movimento:
            writer.writerow([
                empresa.cnpj,
                empresa.nome or "",
                empresa.qtd_notas_emitidas,
                empresa.qtd_notas_recebidas,
                0,
                "Sem movimento"
            ])
        
        csv_content = output.getvalue()
        output.close()
        
        # Retorna como resposta de download
        from fastapi.responses import Response
        filename = f"resumo_execucoes_{competencia or 'todas'}_{datetime.now().strftime('%Y%m%d_%H%M%S')}.csv"
        
        return Response(
            content=csv_content,
            media_type="text/csv",
            headers={
                "Content-Disposition": f"attachment; filename={filename}"
            }
        )
        
    except Exception as e:
        logger.error(f"Erro ao gerar CSV de resumo: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Erro ao gerar CSV: {str(e)}"
        )

File: Backend/src/routers/relatorios.ts
/**
 * Router de relatórios de execuções.
 * Rotas de relatórios.
 */
import { Router, Request, Response } from 'express';
import { getLogger } from '../infrastructure/logger';
import { prisma } from '../db/client';
import * as empresasRepo from '../repositories/empresas';

const logger = getLogger('relatorios');
const router = Router();

interface EmpresaResumo {
  cnpj: string;
  nome: string | null;
  qtd_notas_emitidas: number;
  qtd_notas_recebidas: number;
}

interface ResumoExecucoes {
  competencia: string | null;
  total_empresas: number;
  com_movimento: number;
  sem_movimento: number;
  empresas_com_movimento: EmpresaResumo[];
  empresas_sem_movimento: EmpresaResumo[];
}

// GET /execucoes/resumo - Resumo das execuções
router.get('/execucoes/resumo', async (req: Request, res: Response) => {
  try {
    const competencia = (req.query.competencia as string) || null;
    const statusFiltro = (req.query.status_filtro as string) || 'concluido';

    const where: { status: string; periodoInicio?: { contains: string } } = {
      status: statusFiltro,
    };

    if (competencia && competencia.length === 6 && /^\d+$/.test(competencia)) {
      const mes = competencia.slice(0, 2);
      const ano = competencia.slice(2, 6);
      where.periodoInicio = { contains: `/${mes}/${ano}` };
    }

    const execucoes = await prisma.execucao.findMany({
      where,
      orderBy: { createdAt: 'desc' },
    });

    const empresasComMovimento: EmpresaResumo[] = [];
    const empresasSemMovimento: EmpresaResumo[] = [];

    for (const exec of execucoes) {
      const qtdEmitidas = exec.qtdNotasEmitidas ?? 0;
      const qtdRecebidas = exec.qtdNotasRecebidas ?? 0;

      let nomeEmpresa: string | null = null;
      if (exec.empresaId) {
        try {
          const empresa = await empresasRepo.obterEmpresaPorId(exec.empresaId);
          nomeEmpresa = empresa?.razaoSocial ?? null;
        } catch {
          /* ignore */
        }
      }

      const resumo: EmpresaResumo = {
        cnpj: exec.cnpj ?? String(exec.empresaId),
        nome: nomeEmpresa,
        qtd_notas_emitidas: qtdEmitidas,
        qtd_notas_recebidas: qtdRecebidas,
      };

      if (qtdEmitidas > 0 || qtdRecebidas > 0) {
        empresasComMovimento.push(resumo);
      } else {
        empresasSemMovimento.push(resumo);
      }
    }

    const resultado: ResumoExecucoes = {
      competencia: competencia ?? null,
      total_empresas: execucoes.length,
      com_movimento: empresasComMovimento.length,
      sem_movimento: empresasSemMovimento.length,
      empresas_com_movimento: empresasComMovimento,
      empresas_sem_movimento: empresasSemMovimento,
    };

    res.json(resultado);
  } catch (error) {
    logger.error({ err: error }, 'Erro ao gerar resumo de execuções');
    res.status(500).json({ detail: 'Erro ao gerar resumo de execuções' });
  }
});

// GET /execucoes/resumo/csv - Download CSV do resumo
router.get('/execucoes/resumo/csv', async (req: Request, res: Response) => {
  try {
    const competencia = (req.query.competencia as string) || null;
    const statusFiltro = (req.query.status_filtro as string) || 'concluido';

    const where: { status: string; periodoInicio?: { contains: string } } = {
      status: statusFiltro,
    };

    if (competencia && competencia.length === 6 && /^\d+$/.test(competencia)) {
      const mes = competencia.slice(0, 2);
      const ano = competencia.slice(2, 6);
      where.periodoInicio = { contains: `/${mes}/${ano}` };
    }

    const execucoes = await prisma.execucao.findMany({
      where,
      orderBy: { createdAt: 'desc' },
    });

    const empresasComMovimento: EmpresaResumo[] = [];
    const empresasSemMovimento: EmpresaResumo[] = [];

    for (const exec of execucoes) {
      const qtdEmitidas = exec.qtdNotasEmitidas ?? 0;
      const qtdRecebidas = exec.qtdNotasRecebidas ?? 0;

      let nomeEmpresa: string | null = null;
      if (exec.empresaId) {
        try {
          const empresa = await empresasRepo.obterEmpresaPorId(exec.empresaId);
          nomeEmpresa = empresa?.razaoSocial ?? null;
        } catch {
          /* ignore */
        }
      }

      const resumo: EmpresaResumo = {
        cnpj: exec.cnpj ?? String(exec.empresaId),
        nome: nomeEmpresa,
        qtd_notas_emitidas: qtdEmitidas,
        qtd_notas_recebidas: qtdRecebidas,
      };

      if (qtdEmitidas > 0 || qtdRecebidas > 0) {
        empresasComMovimento.push(resumo);
      } else {
        empresasSemMovimento.push(resumo);
      }
    }

    const linhas: string[] = [];
    const enc = (v: string | number) => `"${String(v).replace(/"/g, '""')}"`;
    linhas.push(['CNPJ', 'Nome', 'Total Notas Emitidas', 'Total Notas Recebidas', 'Total Notas', 'Status'].map(enc).join(','));

    for (const emp of empresasComMovimento) {
      const total = emp.qtd_notas_emitidas + emp.qtd_notas_recebidas;
      linhas.push([emp.cnpj, emp.nome ?? '', emp.qtd_notas_emitidas, emp.qtd_notas_recebidas, total, 'Com movimento'].map(enc).join(','));
    }
    for (const emp of empresasSemMovimento) {
      linhas.push([emp.cnpj, emp.nome ?? '', emp.qtd_notas_emitidas, emp.qtd_notas_recebidas, 0, 'Sem movimento'].map(enc).join(','));
    }

    const csv = linhas.join('\n');
    const filename = `resumo_execucoes_${competencia ?? 'todas'}_${new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19)}.csv`;

    res.setHeader('Content-Type', 'text/csv');
    res.setHeader('Content-Disposition', `attachment; filename="${filename}"`);
    res.send(csv);
  } catch (error) {
    logger.error({ err: error }, 'Erro ao gerar CSV de resumo');
    res.status(500).json({ detail: 'Erro ao gerar CSV de resumo' });
  }
});

export default router;
